### ISSUE TITLE
Chess positive interoperability tracker

### ISSUE THREAD

**Issue by alltheseas** [nostrability/nostrability#70]

**Comment by theborakompanioni** [comment]
> @theborakompanioni which other apps have expressed interest in adopting NIP-64/chess?

Non explicitly expressed interest, but there are issues open on a few selected clients:
- amethyst: https://github.com/vitorpamplona/amethyst/issues/878
- damus: https://github.com/damus-io/damus/issues/2279
- iris: https://github.com/irislib/iris-messenger/issues/452
- nostrudel: https://github.com/hzrd149/nostrudel/issues/184
- primal: https://github.com/PrimalHQ/primal-web-app/issues/71
- satellite: https://github.com/lovvtide/satellite-web/issues/27

Planning on creating a dedicated client myself.
Should issues be opened on the other clients in the list as well?

**Comment by alltheseas** [comment]
@theborakompanioni lets just add the nostr apps which specifically either

1) implemented nip-64, or
2) agreed to implementing nip-64 in the future

Once we have these we will get uncle @rockstardev to use his influencer authority over nostp app devs, we will build a critical mass, and we will then make a case to convert Lichess over to nostr.

what is your app name?

**Comment by alltheseas** [comment]
@theborakompanioni what is your npub

**Comment by alltheseas** [comment]
I am advocating for this NIP https://njump.me/nevent1qqs27kefhjee7jeg5ujtsq7ekyza9shgwlarp83p4skrz4pax7f9fwgpp4mhxue69uhkummn9ekx7mqpr3mhxue69uhkummnw3ezucnfw33k76twv4ezuum0vd5kzmqpzemhxue69uhkzarvv9ejumn0wd68ytnvv9hxgqgcwaehxw309aex2mrp0yh8xmn0wf6zuum0vd5kzmqk88y6g as a venue to bring lichess active users to nostr

**Comment by alltheseas** [comment]
@mikedilger chess app tinkering:

https://njump.me/note1xrymh37r7gqdg798dvsc7kd04fawz0haumxy8247sqp3d3m52chslfkcfs

> Strange coincidence.  I wrote a chess app on the weekend a few weeks ago because I wanted to practice a bunch of Caro-Kann lines (I'm an aggressive player and I wanted to try to play defensive instead to expand my reperitoire).

> The game state library imports/exports FEN or a series of algebraic moves. It uses bitboards and verifies legal moves, 'in check', etc, including en passant, castle rights, halfmove clock, etc.. the only major thing it doesn't do yet is determine checkmate (I just hadn't gotten that far in a weekend). 

> I got it plugged into stockfish and I can play against stockfish in a terminal. I didn't even get started on the gui.

> Then I got it working beyind PyChess as an engine. The engine makes any of the "likely" Caro-Kann opponent moves and then I have to make the correct Caro-Kann response.  This way I could practice those lines.  If it gets into a state I didn't prepare for, it makes an illegal move to end the sequence.

> Anyhow.... 
Even though I was able to write that rather quickly, I don't expect other nostr developers to write a chess engine for NIP-64.  However, in order to display the chessboard from PGN notation, they will need to use a chess engine.... to either write one or to link to one... that makes each move specified one after another, each one validated, and then you end up with a board state.

> It would be far easier for a nostr client to take a FEN and display the board state, because that is what a FEN is. Then nostr clients wouldn't need a chess engine inside of them.

**Comment by theborakompanioni** [comment]
> @theborakompanioni what is your npub

I do not maintain long living npubs.

> what is your app name?

Jester, but it was written long ago, does not leverage NIP-64 and needs a lot of love. Also have plans for other chess related apps.


> @mikedilger chess app tinkering:
> 
> https://njump.me/note1xrymh37r7gqdg798dvsc7kd04fawz0haumxy8247sqp3d3m52chslfkcfs
> 
> > It would be far easier for a nostr client to take a FEN and display the board state, because that is what a FEN is. Then nostr clients wouldn't need a chess engine inside of them.

That sounds reasonable at first and might even be true (everything has up- and downsides), but you can always leverage the `FEN` tag in a PGN database, e.g. 

```pgn
[Event "Puzzle 00008"]
[Site "https://lichess.org/787zsVup/black#47"]
[Date "????.??.??"]
[Round "-"]
[White "?"]
[Black "?"]
[Result "*"]
[FEN "r6k/pp2r2p/4Rp1Q/3p4/8/1N1P2R1/PqP2bPP/7K b - - 0 24"]

*
```

Most clients would even accept a short version:

```pgn
[FEN "r6k/pp2r2p/4Rp1Q/3p4/8/1N1P2R1/PqP2bPP/7K b - - 0 24"]

*
```

**Comment by mikedilger** [comment]
> you can always leverage the `FEN` tag in a PGN database

If it has one.  None of the examples in NIP-64 have one.

**Comment by mikedilger** [comment]
Just for clarity, when I said "chess engine" in my post I meant game-state code (not an engine that figures out the best move like stockfish does).

PGN uses algebraic move notation like "e4".  To know what "e4" means you need to know whose turn it is, where the pawns are on the e file, and maybe even check that it is legal to move the pawn from e2 to e4 (nothing is on e3, it doesn't put the king in check, etc).  Also for say "Nc3" you have to "find a knight of the player character which can legally move to c3 and hope there is just one".  While humans like this notation, it is complex for a nostr client to work them out.

There is a much more direct notation that stockfish uses "e2e4" which just specifies the source square and the destination square (and maybe a promotion piece after). Any idiot could code that without needing to understand how the game of chess works (and it is also far faster than algebraic as nothing needs to be worked out, which is why stockfish uses it).

So nostr clients could display and allow users to play chess without understanding chess at all if we sent a series of (source-dest-promotion) moves, and some DVM acted as a referee.

But that's not NIP-64, that is playing chess over nostr.

**Comment by mikedilger** [comment]
https://github.com/nostr-protocol/nips/issues/1453

**Comment by theborakompanioni** [comment]
> > you can always leverage the `FEN` tag in a PGN database
> 
> If it has one. None of the examples in NIP-64 have one.

True, but it really just describes the interpretation of PGN format in clients (i.e. display them as chessboards).
The intention was to keep the NIP spec as simple and concise as possible. It cannot have all the possible combinations listed as examples, as the PGN format in itself is not trivial.

However, you are right, that NIP-64 does not tell anything about chess rules, and that is a Good Thing™.
Libraries do not need to include move validation to display chessboards, e.g. https://github.com/lichess-org/chessground

> So nostr clients could display and allow users to play chess without understanding chess at all if we sent a series of (source-dest-promotion) moves, and some DVM acted as a referee.
> 
> But that's not NIP-64, that is playing chess over nostr.

Yes, I think that is the main takeaway: If you create a chess app you can use any format you want, whenever the game finishes, publish the result as NIP-64 event. Personally, at the time of writing, I would still use NIP-64 as the main transport for game state and use expiration tags and other nostr features to keep the footprint on relays smaller. 

> [nostr-protocol/nips#1453](https://github.com/nostr-protocol/nips/issues/1453)

Nice. I really love to see more exploration of this and welcome any and each approach. There are some many things to consider, like time control, chess variants, custom rules, etc. If it becomes a standard in nostr, it should support a wide range of everything that is chess. Additionally, there are tricks things in "standard chess" like "offer draw and agreeing on a draw", "offer draw and withdrawing the offer", "claiming or not claiming a draw after threefold repetition", etc.

Having said that, I want to point out that it is not trivial and there are so many opinions on that with all sort of (good!) arguments on every side.

**Comment by theborakompanioni** [comment]
> Gossip won't do it so you can mark it off on your list. I think chess should probably be run in a custom chess client.

Perfectly fine to not implement it. :raised_hands: 
Care to elaborate the reasoning behind this decision? I am interested. 

However, since it is PGN it is human readable I think text-representation is okay (and user can always import it elsewhere if they care about the specific game).

**Comment by mikedilger** [comment]
> Perfectly fine to not implement it. 🙌 Care to elaborate the reasoning behind this decision? I am interested.

Mostly because I don't want to maintain a giant client that tries to do "everything nostr".  Or maybe for the same reason that I don't want my refrigerator to also be my pocket knife, or my surfboard to have wheels.  They just don't go together IMHO. Gossip is for microblogging like twitter.  Gossip won't even have chat.

**Comment by alltheseas** [comment]
https://crates.io/crates/rooky-core

```
Rooky is a protocol for sharing chess games and annotations using the Nostr protocol. It allows users to publish, share, and view chess games in a decentralized and secure manner.

Rooky extends NIP-64, using PGN format for the chess data interoperability.

Features
Rooky Core
The main library rooky-core provides the common structures and definition to create Nostr messages with PGN content. It uses the shakmaty and pgn-reader crates to provide an easy data structure to parse, manage, and serialize chess games.

ChessBoard JS
This include WebAssembly bindings for the ChessboardJS library, used to display interactive chess boards on the application. WebAssembly allows for high-performance rendering of chess boards in the browser.

External
This defines a set of helper functions to interact with external APIs from sites like lichess and Chess.com. Uses browser native code to create efficient streams of games to avoid overhead on the connections.

Technologies
Rooky is built on Rust and leverages the nostro2 to create easy Nostr structures that can be used with both software signers and signing extensions.

The RookyGame structures support full serde serializations to ensure easy compatibility with other systems and libraries.

The RookyGame struct also uses the shakmaty crate to provide a rich set of chess functionalities, including checking the legality of the moves before applying them, and managing and interacting with the game state.

The RookyGame can be converted to a Nostr note for interaction with the network, or can produce the raw PGN if needed, eg for text downloads.
```
https://github.com/satsangatech/rooky-chess-protocol

**Comment by alltheseas** [comment]
https://github.com/AsaiToshiya/brostr allows for viewing nip-64 chess
